add support for css `@property`

z4o4z opened the pull request:
This PR adds support for [CSS @property](https://developer.mozilla.org/en-US/docs/Web/CSS/@property) feature. 

- `createVar` - overloaded to allow a `@property` definition, returns property var format `var(--propertyName)`
- `createGlobalVar` - same as `createVar` but with a global (non-scoped) name

example:
```ts
import { style, createVar, keyframes, fallbackVar } from '@vanilla-extract/css';

const color = createVar();
const angle = createVar({
  syntax: '<angle>',
  inherits: false,
  initialValue: '0deg',
});

const angleKeyframes = keyframes({
  '100%': {
    vars: {
      [getVarName(angle)]: '360deg',
    }
  },
});

export const root = style({
  background: 'pink',
  color: 'blue',
  padding: '16px',
  transition: `opacity .1s ease, color .1s ease`, // Testing autoprefixer
  backgroundImage: `linear-gradient(${angle}, rgba(153, 70, 198, 0.35) 0%, rgba(28, 56, 240, 0.46) 100%)`,
  animation: `${angleKeyframes} 7s infinite ease-in-out both`,


  ':hover': {
    opacity: 0.8,
    color: color,
  },
  
  vars: {
    [color]: '#fef',
    [angle]: fallbackVar(angle, '138deg'),
  }
});
```

markdalgleish commented:
Thanks for the PR! This looks really great.

To me this is just an extension of our current `createVar` API. The fact that it's natively called `@property` definitely challenges our current naming, but I think for now VE needs to be consistent in calling them "vars". If we rename them to properties, we'd do this across the whole library in a future major version. I think it's okay that internals refer to it as "property" though since this future-proofs the integration package against API changes.

So I think we'd need to make the following changes for now:
- Overload `createVar` so the API is `createVar({ syntax: '<color>', inherits: false, initialValue: '#fff' })`.
- Rename `createGlobalProperty` to `createGlobalVar`.
- Replace `getPropertyName` with `getVarName` (which already exists in our `private` package). This would need to be exposed from `css` and `dynamic` packages since you'd want this at both build time and run time.

z4o4z commented:
@markdalgleish thanks for cleaning this up! is there something I should address before merging?

z4o4z commented:
@markdalgleish any updates here, i'm really waiting for this feature?

erwinsmit commented:
> @markdalgleish any updates here, i'm really waiting for this feature?

This would be a nice addition. Currently I can only use properties using globalStyle with a @ts-ignore. That feels a bit dirty 

```javascript
globalStyle(`@property --headerBackground1`, {
  // @ts-ignore
  syntax: `'<color>'`,
  initialValue: colorPrimitives.neutrals.whites.W100,
  inherits: false,
});
```

8bit-echo commented:
I'd like to bump this and see it land! 🙏  This would be a great feature for our design system to use.

mikaalnaik commented:
Need this as well, any update to this PRs future @markdalgleish?

lukelarsen commented:
I am also needing this in our current library. Any update?

askoufis reviewed (commented):
Hey, I'm keen to help get this PR through the door. For me, there's a few things still missing.

## Testing `@property` CSS output in a fixture

Your example in `examples/next` is somewhat complex, and isn't actually being tested anywhere. We snapshot screenshots and CSS output of each of the `fixture` packages, so maybe adding a simple property in the `themed` fixture would be enough.

## Unit testing CSS output

This is done in `transformCss.test.ts`. You probably just need to add a single test case that uses an object with `type: 'property'`.

## Docs

We tend to be a bit particular about docs, so that _can_ be left to a separate PR. However, you're more than welcome try write them yourself if you want.

askoufis commented:
@z4o4z Hey, are you still interested in getting this PR over the line? If not, I may look at picking it up myself in the future.

z4o4z commented:
hey @askoufis, i've added tests and updated docs. let me know if that looks good for you :) Thanks for helping with the PR!

askoufis reviewed (commented):
Thanks for picking this up again @z4o4z, it's most of the way there. For me the last missing bit is docs for `createGlobalVar`, which should be fairly straightforward. I've also made a comment about `keyframes` below.

Once the last bits are ready there's also some playwright snapshots that need updating.

z4o4z commented on site/docs/api/create-var.md:
```diff
@@ -112,3 +112,59 @@ export const pink = style({
+  '100%': {
+    [getVarName(angle)]: '360deg',
+  },
+});
```
hey @askoufis, thanks for your feedback, this is a really good improvement!

z4o4z commented:
@askoufis addressed the feedback + improved documentation. I found out that parcel doesn't support `@property` and all parcel-related playwright tests are failing now, not sure what to do with it...

z4o4z commented:
> @askoufis addressed the feedback + improved documentation. I found out that parcel doesn't support `@property` and all parcel-related playwright tests are failing now, not sure what to do with it...

looks like it's not the parcel issue itself, parcel uses `lightningcss` to build CSS, and `lightningcss` doesn't support `@property` keyword ([playground](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Atrue%2C%22customMedia%22%3Atrue%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A6225920%7D%2C%22include%22%3A0%2C%22exclude%22%3A0%2C%22source%22%3A%22%40property%20--blankVar1__jteyb14%20%7B%5Cn%20%20syntax%3A%20%5C%22%3Cnumber%3E%5C%22%3B%5Cn%20%20inherits%3A%20false%3B%5Cn%7D%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D)) :(

z4o4z commented on packages/css/src/vars.ts:
```diff
@@ -25,7 +25,7 @@ type VarDeclaration = {
   syntax: `"${Array.isArray(syntax) ? syntax.join(' | ') : syntax}"`,
   inherits: inherits ? 'true' : 'false',
-  initialValue,
+  ...(initialValue != null && { initialValue }),
```
need this to don't have `initial-value: undefined` in css

z4o4z commented on site/docs/api/create-var.md:
````diff
@@ -137,34 +137,6 @@ export const pink = style({
     [accentVar]: 'pink'
   }
 });
-``` 
````
moved to `keyframes` doc

z4o4z commented on fixtures/themed/src/styles.css.ts:
```diff
@@ -67,7 +67,6 @@ globalStyle(`body ${iDunno}:after`, {
 const blankVar1 = createVar({
   syntax: '<number>',
   inherits: false,
-  initialValue: '0',
```
removed initial value to make fallback function `fallbackVar(blankVar1, '0.5')` work

z4o4z commented:
looks like it wasn't an issue in the parcel, apparently `initial-value` is required for non `*` syntax, so it was failing with not really useful error. So i've made `initialValue` required on type level for non `*` syntax.

askoufis reviewed (commented):
Thanks for the quick updates. Just some more minor comments from me.

Also realised there's no changesets for the features. ~We'll need one for `createVar` and `createGlobalVar`, as well as a separate one for adding `vars` support to `keyframes`.~ There's actually a few more changes that need changeesets. I'll add them.

askoufis commented:
@markdalgleish Would you mind giving this another review when you've got some time?

askoufis commented:
Looks like linting is also failing, specifically `prettier`.

askoufis commented on packages/babel-plugin-debug-ids/src/index.ts:
```diff
@@ -31,7 +31,11 @@ const debuggableFunctionConfig = {
   },
   createVar: {
-    maxParams: 1,
+    maxParams: 2,
+    hasDebugId: ({ arguments: args }) => {
+      const previousArg = args[args.length - 1];
+      return t.isStringLiteral(previousArg) || t.isTemplateLiteral(previousArg);
+    },
```
A test case should be added for this functionality.

askoufis commented on packages/css/src/transformCss.ts:
```diff
@@ -150,10 +153,17 @@ class Stylesheet {
       root.rule = Object.fromEntries(
         Object.entries(root.rule).map(([keyframe, rule]) => {
-          return [keyframe, this.transformProperties(rule)];
+          return [keyframe, this.transformVars(this.transformProperties(rule))];
```
This needs a test case too.

z4o4z commented on packages/css/src/vars.ts:
```diff
@@ -9,20 +11,62 @@ import cssesc from 'cssesc';
   return `var(--${cssVarName})` as const;
 }
 
+export function createGlobalVar(name: string): CSSVarFunction
```
i guess it's better to have api to create a global var (`createGlobalVar('name')`) instead of manually typing `var(--name)`

z4o4z commented:
@askoufis thanks for the feedback! addressed everything

z4o4z commented:
hey @askoufis @markdalgleish any chance we can merge it soon?

Vishtar commented:
Bump.

ShayDavidson commented:
bump! I just need this one as well ;)

askoufis commented on packages/css/src/types.ts:
```diff
@@ -155,3 +162,22 @@ export type ThemeVars<ThemeContract extends NullableTokens> = MapLeafNodes<
 
 export type ComplexStyleRule = StyleRule | Array<StyleRule | ClassNames>;
+
+export type PropertySyntax =
+  | '<length>'
+  | '<number>'
+  | '<percentage>'
+  | '<length-percentage>'
+  | '<color>'
+  | '<image>'
+  | '<url>'
+  | '<integer>'
+  | '<angle>'
+  | '<time>'
+  | '<resolution>'
+  | '<transform-function>'
+  | '<custom-ident>'
+  | '<transform-list>'
+  | '*'
+  // needs this to make TS suggestions work
+  | (string & Record<never, never>);
```
```suggestion
type _PropertySyntax =
  | '<angle>'
  | '<color>'
  | '<custom-ident>'
  | '<image>'
  | '<integer>'
  | '<length-percentage>'
  | '<length>'
  | '<number>'
  | '<percentage>'
  | '<resolution>'
  | '<string>'
  | '<time>'
  | '<transform-function>'
  | '<transform-list>'
  | '<url>'
  | '*';

/** Suggests elements of the provided string union `Suggestions` while still accepting any string */
type LooseAutocomplete<Suggestions extends string> =
  | Suggestions
  | Omit<string, Suggestions>;

export type PropertySyntax = LooseAutocomplete<_PropertySyntax>;

```

askoufis commented on .changeset/angry-rules-tease.md:
```diff
@@ -0,0 +1,5 @@
+"@vanilla-extract/babel-plugin-debug-ids": minor
+---
+
+Add support for `createVar` calls with property definitions
```
```suggestion
Add support for `createVar` calls that declare `@property` rules
```

askoufis commented on .changeset/modern-cobras-sort.md:
```diff
@@ -0,0 +1,28 @@
+"@vanilla-extract/css": minor
+---
+
+Add support for a `vars` property to steps within `keyframes` declarations
```
```suggestion
`keyframes`: Add support for a `vars` property to steps within `keyframes` declarations
```

askoufis commented on .changeset/modern-cobras-sort.md:
````diff
@@ -0,0 +1,28 @@
+Example usage:
+
+```ts
+const angle = createVar({
+  syntax: '<angle>',
+  inherits: false,
+  initialValue: '0deg'
+});
+
+const angleKeyframes = keyframes({
````
```suggestion
import { createVar, keyframes } from '@vanilla-extract/css';

const angle = createVar({
  syntax: '<angle>',
  inherits: false,
  initialValue: '0deg'
});

export const angleKeyframes = keyframes({
```

askoufis commented on .changeset/poor-mirrors-care.md:
```diff
@@ -0,0 +1,28 @@
+"@vanilla-extract/babel-plugin-debug-ids": minor
+---
+
+Add support for property definitions to `createVar`
```
```suggestion
`createVar`: Add support for defining [`@property`] rules
```

askoufis commented on .changeset/poor-mirrors-care.md:
```diff
@@ -0,0 +1,28 @@
+---
+
+Add support for property definitions to `createVar`
+
+When a property definition is provided to `createVar`, a `@property` CSS rule will be generated.
```
```suggestion
```

askoufis commented on .changeset/poor-mirrors-care.md:
````diff
@@ -0,0 +1,28 @@
+```ts
+import { createVar } from '@vanilla-extract/css';
+
+export myVar = createVar({
````
```suggestion
export const myVar = createVar({
```

askoufis commented on .changeset/poor-mirrors-care.md:
````diff
@@ -0,0 +1,28 @@
+  inherits: false;
+  initial-value: 0.5;
+}
+```
````
````suggestion
```

[`@property`]: https://developer.mozilla.org/en-US/docs/Web/CSS/@property
````

askoufis commented on site/docs/api/create-var.md:
````diff
@@ -112,3 +112,31 @@ export const pink = style({
 
 CSS variables can also be assigned dynamically using APIs in [the `@vanilla-extract/dynamic` package](/documentation/packages/dynamic).
+
+## CSS Properties
+
+You can also create typed css [@property](https://developer.mozilla.org/en-US/docs/Web/CSS/@property) using `createVar`:
+
+```ts compiled
+// accent.css.ts
+
+import { createVar, style } from '@vanilla-extract/css';
+
+export const accentVar = createVar({
+  syntax: '<color>',
+  inherits: false,
+  initialValue: 'blue'
+});
+```
+
+CSS properties created with `createVar` can be used in the same way as a regular css variable:
+
+```ts compiled
+export const pink = style({
+  vars: {
+    [accentVar]: 'pink'
+  }
+});
+```
+
+The main benefit of using `createVar` to create typed css properties is that they can be [animated via keyframes](/documentation/api/keyframes#animating-variables).
````
```suggestion
## @property rules

[@property] rules can also be created using `createVar`.
CSS variables with @property rules are used in the same way as regular CSS variables:

```ts compiled
// accent.css.ts
import { createVar, style } from '@vanilla-extract/css';

export const accentVar = createVar({
  syntax: '<color>',
  inherits: false,
  initialValue: 'blue'
});

export const pink = style({
  vars: {
    [accentVar]: 'pink'
  }
});
```

[@property]: https://developer.mozilla.org/en-US/docs/Web/CSS/@property
```

askoufis commented on site/docs/global-api/create-global-var.md:
```diff
@@ -0,0 +1,44 @@
+
+# createGlobalVar
+
+Similar to `createVar`, `createGlobalVar` creates a single global CSS Variable reference.
```
```suggestion
Similar to [`createVar`], `createGlobalVar` creates a global CSS variable reference:
```

askoufis commented on site/docs/global-api/create-global-var.md:
````diff
@@ -0,0 +1,44 @@
+});
+```
+
+It's also possible to define typed CSS properties via `createGlobalVar`:
````
```suggestion
[@property] rules can also be created using `createGlobalVar`:
```

askoufis commented on site/docs/global-api/create-global-var.md:
````diff
@@ -0,0 +1,44 @@
+export const content = style({
+  opacity: opacityVar
+});
+```
````
````suggestion
```

[`createVar`]: /documentation/api/create-var
[@property]: https://developer.mozilla.org/en-US/docs/Web/CSS/@property
````

askoufis commented on site/docs/api/keyframes.md:
````diff
@@ -27,3 +27,38 @@ export const spinAgain = style({
 });
 ```
+
+## Animating variables
+
+It is also possible to animate [typed CSS property](/documentation/api/create-var#css-property) using keyframes:
+
+```ts compiled
+const angle = createVar({
+  syntax: '<angle>',
+  inherits: false,
+  initialValue: '0deg'
+});
+
````
````suggestion
## Animating CSS variables

CSS variables can be animated by setting their values within the `vars` property of a keyframe step:

```ts compiled
// animation.css.ts

import {
  createVar,
  fallbackVar,
  keyframes,
  style
} from '@vanilla-extract/css';

````

askoufis reviewed (commented):
I had a small handful of tweaks I wanted to make. Was going to push them to your fork but I don't have permission to do so (I think because it's the default branch of a fork). I've instead made suggestions for the changes and will commit them.

askoufis reviewed (approved):
~Seems like I can't run CI on your `master` fork. I will merge this and fix up anything that breaks in a subsequent PR.~ Really appreciate your work and everyone's patience on this feature. Thank you!

askoufis commented on packages/rollup-plugin/test/__snapshots__/rollup-plugin.test.ts.snap:
```diff
@@ -213,8 +218,8 @@ export { shadow };
     "import './../assets/src/shared.css.ts.vanilla-G_Gyt4-e.css';
-import './../assets/src/themes.css.ts.vanilla-s9rcEmBH.css';
-import './../assets/src/styles.css.ts.vanilla-C7NI1Ovc.css';
+import './../assets/src/themes.css.ts.vanilla-DMMXGwB8.css';
```
```suggestion
import './../assets/src/themes.css.ts.vanilla-s9rcEmBH.css';
```

askoufis commented:
This has been released in [`@vanilla-extract/css@1.17.0`](https://github.com/vanilla-extract-css/vanilla-extract/releases/tag/%40vanilla-extract/css%401.17.0).

ShayDavidson commented:
yay!
